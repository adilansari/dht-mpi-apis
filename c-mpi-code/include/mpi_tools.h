
#ifndef MPI_TOOLS
#define MPI_TOOLS

#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <mpi.h>

#ifdef ENABLE_DEBUG
#ifdef HAVE_EXEC_INFO_H
#include <execinfo.h>
#endif
#endif

#include <gossip.h>

extern int mpi_rank;
extern int mpi_size;

extern int debug_rank;

extern FILE* debug_file;

/**
   Quick setup
 */
void whoami(void);

char* heap(char* s);
char* iheap(int i);
char* xheap(int i);

/**
   Execute code line or block in rank order.
*/
#define sequential(code)                                                \
  {                                                                     \
    int sequential_i;                                                   \
    for (sequential_i = 0; sequential_i < mpi_size;                     \
         sequential_i++)                                                \
    {                                                                   \
      if (mpi_rank == sequential_i)                                     \
      {                                                                 \
        code;                                                           \
      }                                                                 \
      MPI_Barrier(MPI_COMM_WORLD);                                      \
    }                                                                   \
  }

void printbinary(unsigned int number);
void printdata(char* tag, char* data, int length);
void fprintdata(FILE* file, char* tag, char* data, int length);
int  flip2(int i);
unsigned int ilog2(unsigned int j);
unsigned int isqrt(unsigned int j);

/**
   Return a random number in [0,n).
*/
int rand_lt(int n);

/**
   Return a random double in [0,n).
*/
double random_double(int n);

/**
   Concatenate char* s at char* p, update p to the end.
*/
#define cat(p, s) p += sprintf(p, "%s", s);

#ifdef ENABLE_DEBUG
// #warning "ENABLE_DEBUG"
#define DEBUG(x) x
#else
#define DEBUG(x)
#endif

#ifdef ENABLE_DEBUG
/**
   Dump the stack and exit.
*/
void
dump_stack(void);
#endif

#define OK NOTE("ok")

void setup_debug_file(void);

/**
   Simplify debugging statements.
 */
void note(char* msg);
void note_b(char* msg, int b);
void note_d(char* msg, double d);
void note_i(char* msg, int i);
void note_ii(char* msg, int i, int j);
void note_s(char* msg, char* s);
void note_ss(char* msg, char* s1, char* s2);
void note_p(char* msg, void* p);
void note_u(char* msg, unsigned int u);
void note_x(char* msg, int i);
void note_xi(char* msg, int i, int j);
void note_sx(char* msg, char* s, int i);
void note_sxi(char* msg, char* s, int x, int i);
// void note_ssxi(char* msg, char* s, int x, int i);

void show_b(char* tag, int b);
void show_d(char* tag, double d);
void show_i(char* tag, int i);
void show_ii(char* tag_i, int i, char* tag_j, int j);
void show_l(char* tag, long l);
void show_s(char* tag, char* s);
void show_u(char* tag, unsigned int u);
void show_x(char* tag, int x);
void show_p(char* tag, void* p);

void  timestamp(char* label, char* msg);
char* timestring(char* label, char* msg);

#define NOTE(m)                DEBUG(note(m))
#define NOTE_B(m,b)            DEBUG(note_b(m,b))
#define NOTE_D(m,d)            DEBUG(note_d(m,d))
#define NOTE_I(m,i)            DEBUG(note_i(m,i))
#define NOTE_II(m,i,j)         DEBUG(note_ii(m,i,j))
#define NOTE_U(m,u)            DEBUG(note_u(m,u))
#define NOTE_P(m,s)            DEBUG(note_p(m,))
#define NOTE_S(m,s)            DEBUG(note_s(m,s))
#define NOTE_SX(m,s,i)         DEBUG(note_sx(m,s,i))
#define NOTE_SXI(m,s,x,i)      DEBUG(note_sxi(m,s,x,i))
// #define NOTE_SSXI(m,s1,s2,x,i) DEBUG(note_ssxi(m,s1,s2,x,i))
#define NOTE_X(m,i)            DEBUG(note_x(m,i))
#define NOTE_XI(m,i,j)         DEBUG(note_xi(m,i,j))

#define NOTE_F DEBUG(fprintf(debug_file, "[%i] %s()... \n",     \
                             debug_rank, __func__);             \
                     fflush(debug_file);)
#define DONE DEBUG(fprintf(debug_file, "[%i] %s() done. \n",     \
                          debug_rank, __func__);        \
                   fflush(debug_file);)
#define NOTE_FI(i)                              \
  DEBUG(fprintf(debug_file, "[%i] %s(%i)... \n",             \
               debug_rank, __func__, i);        \
        fflush(stdout);)
#define NOTE_FX(i) DEBUG(fprintf(debug_file, "[%i] %s(%X)... \n",            \
                                debug_rank, __func__, i);       \
                         fflush(stdout);)
#define NOTE_FIS(i,s) DEBUG(fprintf(debug_file, "[%i] %s(%i,%s)... \n",      \
                                   debug_rank, __func__, i, s); \
                            fflush(stdout);)
#define NOTE_FXIS(x,i,s)                        \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%i,%s)... \n",       \
               debug_rank, __func__, x, i, s);  \
        fflush(stdout);)
#define NOTE_FXI(i,j)                           \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%i)... \n",          \
               debug_rank, __func__, i, j);     \
        fflush(stdout);)
#define NOTE_FXS(i,s)                           \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%s)... \n",          \
               debug_rank, __func__, i, s);     \
        fflush(stdout););
#define NOTE_FXX(i,j) DEBUG(fprintf(debug_file, "[%i] %s(%X,%X)... \n",      \
                                   debug_rank, __func__, i, j); \
                            fflush(stdout);)
#define NOTE_FS(s) DEBUG(fprintf(debug_file, "[%i] %s(%s)... \n",            \
                                debug_rank, __func__, s);       \
                         fflush(stdout);)
#define NOTE_FSS(s1,s2)                         \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s)... \n",          \
               debug_rank, __func__, s1, s2);   \
        fflush(stdout);)
#define NOTE_FSSS(s1,s2,s3)                     \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s,%s)... \n",       \
               debug_rank, __func__,            \
               s1, s2, s3);                     \
        fflush(stdout);)
#define NOTE_FSSSI(s1,s2,s3,i)                          \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s,%s,%i)... \n",            \
               debug_rank, __func__, s1, s2, s3, i);    \
        fflush(stdout);)

///// Simple variable debugging...
#define SHOW_B(b)    DEBUG(show_b(#b, b))
#define SHOW_D(d)    DEBUG(show_d(#d, d))
#define show_i(i)    note_i(#i ": ", i)
#define SHOW_I(i)    DEBUG(show_i(i))
#define SHOW_II(i,j) DEBUG(show_ii(#i, i, #j, j))
#define SHOW_L(l)    DEBUG(show_l(#l, l))
#define SHOW_U(u)    DEBUG(show_u(#u, u))
#define show_s(s)    note_s(#s ": ", s)
#define SHOW_S(s)    DEBUG(show_s(s))
#define SHOW_X(x)    DEBUG(show_x(#x, x))
#define SHOW_P(p)    DEBUG(show_p(#p, p))
#define SHOW         SHOW_S

//// Functions...
#define show_fi(i)                              \
       fprintf(debug_file, "[%i] %s(%s=%i)... \n",          \
               debug_rank, __func__, #i, i);    \
        fflush(stdout);
#define SHOW_FI(i) DEBUG(show_fi(i))
#define SHOW_FIS(i,s)                                   \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s)... \n",            \
               debug_rank, __func__, #i, i, #s, s);     \
        fflush(stdout);)
#define SHOW_FISS(i,s1,s2)                                      \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s,%s=%s)... \n",              \
               debug_rank, __func__, #i, i, #s1, s1, #s2, s2);  \
        fflush(stdout);)
#define SHOW_FISX(i,s,x)                                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s,%s=%X)... \n",         \
                debug_rank, __func__, #i, i, #s, s, #x, x);             \
        fflush(stdout);)

#define show_fs(s) fprintf(debug_file,           \
                           "[%i] %s(%s=%s)... \n",          \
               debug_rank, __func__, #s, s);    \
        fflush(stdout);
#define SHOW_FS(s)                              \
  DEBUG(show_fs(s))

#define SHOW_FX(x)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X)... \n",            \
               debug_rank, __func__, #x, x);      \
        fflush(stdout);)
#define SHOW_FII(i1,i2)                                     \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%i)... \n",                \
               debug_rank, __func__, #i1, i1, #i2, i2);     \
        fflush(stdout);)

#define show_fiii(i1,i2,i3)                              \
      fprintf(debug_file, "[%i] %s(%s=%i,%s=%i,%s=%i)... \n",  \
              debug_rank, __func__, #i1, i1, #i2, i2, #i3, i3);   \
        fflush(stdout);
#define SHOW_FIII(i1,i2,i3) DEBUG(show_fiii(i1,i2,i3))

#define show_fsi(s,i)  \
      fprintf(debug_file, "[%i] %s(%s=%s,%s=%i)... \n",  \
                debug_rank, __func__, #s, s, #i, i);     \
        fflush(stdout);
#define SHOW_FSI(s,i) DEBUG(show_fsi(s,i))

#define show_fsii(s,i1,i2)                                     \
  fprintf(debug_file, "[%i] %s(%s=%s,%s=%i,%s=%i)... \n",      \
              debug_rank, __func__, #s, s, #i1, i1, #i2, i2);  \
  fflush(stdout);
#define SHOW_FSII(s,i1,i2) DEBUG(show_fsii(s,i1,i2))

#define SHOW_FXI(x,i)                                   \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i)... \n",            \
               debug_rank, __func__, #x, x, #i, i);     \
        fflush(stdout);)

#define SHOW_FXIS(x,i,s)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i,%s=%s)... \n",      \
               debug_rank, __func__,                    \
               #x, x, #i, i, #s, s);                    \
        fflush(stdout);)

#define SHOW_FXIIS(x,i1,i2,s)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i,%s=%i,%s=%s)... \n",      \
               debug_rank, __func__,                    \
               #x, x, #i1, i1, #i2, i2, #s, s);                    \
        fflush(stdout);)

#define SHOW_FSS(s1,s2)                                 \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s,%s=%s)... \n",            \
               debug_rank, __func__, #s1, s1, #s2, s2); \
        fflush(stdout);)
#define SHOW_FSSI(s1,s2,i)                                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s,%s=%s,%s=%i)... \n",        \
                debug_rank, __func__, #s1, s1, #s2, s2, #i, i);        \
        fflush(stdout);)
#define SHOW_FISSI(i1,s1,s2,i2)                                \
  DEBUG(fprintf(debug_file,                                        \
                "[%i] %s(%s=%i,%s=%s,%s=%s,%s=%i)... \n",    \
                debug_rank, __func__, #i1, i1, #s1, s1,   \
                #s2, s2, #i2, i2); \
        fflush(stdout);)
#define SHOW_FSISSI(s1,i1,s2,s3,i2)                                \
  DEBUG(fprintf(debug_file,                                        \
                "[%i] %s(%s=%s,%s=%i,%s=%s,%s=%s,%s=%i)... \n",    \
                debug_rank, __func__, #s1, s1, #i1, i1, #s2, s2,   \
                #s3, s3, #i2, i2); \
        fflush(stdout);)
#define SHOW_FP(p)                              \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%p)... \n",          \
               debug_rank, __func__, #p, p);    \
        fflush(stdout);)

//// Functions with return values...

#define SHOW_FS_X(s,x)                                                  \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s) -> %X... \n",               \
                debug_rank, __func__, #s, s, x);                        \
        fflush(stdout);)
#define SHOW_FIFS_S(i,f,s1,s2)                                       \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s(%s) -> %s)... \n",     \
                debug_rank, __func__, #i, i, f, s1, s2);             \
        fflush(stdout);)
#define SHOW_FIS_X(i,s,x)                                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s) -> %X... \n",       \
                debug_rank, __func__, #i, i, #s, s, x);               \
        fflush(stdout);)
#define SHOW_FXI_B(x,i,b)                                               \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i) -> %s... \n",         \
                debug_rank, __func__,                                   \
                #x, x, #i, i, b ? "true" : "false");                    \
        fflush(stdout);)


//// Functions of functions...

#define SHOW_FIF(i, f)                               \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s())... \n",           \
               debug_rank, __func__, #i, i, f);       \
        fflush(stdout);)
#define SHOW_FIFS(i, f, s)                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s(%s))... \n",           \
               debug_rank, __func__, #i, i, f, s);      \
        fflush(stdout);)

#define TIMESTAMP(m) note_i(m, time());

#define DUMPSTACK DEBUG(dump_stack());

char* file_string(char* file);

//// Memory usage debugging:
#ifdef DMALLOC
#include <dmalloc.h>
#define DMALLOC_SETUP(x) x
void dmalloc_setup(void);
#else
#define DMALLOC_SETUP(x)
#define dmalloc_setup(x)
#endif

#define DMALLOC_COMPLETE                              \
  DMALLOC_SETUP(                                      \
    {                                                 \
      SHOW_S(dmalloc_logpath);                        \
      dmalloc_shutdown();                             \
      dmalloc_log_unfreed();                          \
    });

/**
   Return a number in [0,n-1].
*/
int roll(int n);

#endif
