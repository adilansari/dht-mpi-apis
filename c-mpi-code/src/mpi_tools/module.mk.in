
DIR := src/mpi_tools

MPITOOLS_SRC :=
MPITOOLS_SRC += $(DIR)/mpi_tools.c
MPITOOLS_SRC += $(DIR)/io_tools.c

MPITOOLS_OBJS = $(patsubst %.c, %.o, $(MPITOOLS_SRC))
