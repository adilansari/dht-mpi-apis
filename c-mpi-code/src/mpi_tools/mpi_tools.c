
#include "mpi_tools.h"

int mpi_size;
int mpi_rank;

int debug_rank = -1;

char output[100];

bool  debug_multiple = false;
FILE* debug_file     = NULL;

void
setup_debug_file(void)
{
  char debug_path[100];

  // debug_multiple = true;

  if (debug_file == NULL)
  {
    if (debug_multiple)
    {
      if (mpi_rank == 0)
        printf("Using multiple debug files \n");
      sprintf(debug_path, "%s%i.out", "debug", debug_rank);
      debug_file = fopen(debug_path, "w");
    }
    else
    {
      if (mpi_rank == 0)
        DEBUG(puts("Debug to stdout"));
      debug_file = stdout;
    }
  }
}

void whoami()
{
  MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);

  if (debug_rank == -1)
    debug_rank = mpi_rank;

  DEBUG(setup_debug_file());
}

/*
void
sync_output(int rank, int size, char* msg)
{
  int max_length;
  int length     = strlen(msg)+1;

  MPI_Reduce(&length, &max_length, 1, MPI_INT, MPI_MAX, 0,
             MPI_COMM_WORLD);

  char *buf;

  if (rank != 0)
  {
    MPI_Send(msg, length, MPI_CHAR, 0, 0, MPI_COMM_WORLD);
  }
  else
  {
    int i;
    MPI_Status status;
    buf = (char*) malloc(sizeof(char)*max_length);

    for (i = 1; i < size; i++)
    {
      MPI_Recv(buf, max_length, MPI_CHAR, i,
               0, MPI_COMM_WORLD, &status);
      printf("%s\n", buf);
    }
  }
}
*/

/**
   Print possibly binary data with nice handling of newline, tab,
   NULL characters, and misc. non-printing characters
 */
void
fprintdata(FILE* file, char* tag, char* data, int length)
{
  int i;
  int c;
  fprintf(file, "%s[", tag);
  for (i = 0; i < length; i++)
  {
    c = data[i];
    if (c == '\0')
      fprintf(file, "\\0");
    else if (c == '\n')
      fprintf(file, "\\n");
    else if (c == '\t')
      fprintf(file, "\\t");
    else if (c >= 32 && c <=126)
      fprintf(file, "%c", c);
    else
      fprintf(file, "\\%i", (int) c);
  }
  fprintf(file, "]");
}

/**
   Print possibly binary data with nice handling of newline, tab,
   NULL characters, and misc. non-printing characters
 */
void
printdata(char* tag, char* data, int length)
{
  fprintdata(stdout, tag, data, length);
}

/**
   Print the bits of a binary number
 */
void
printbinary(unsigned int number)
{
  int x;
  for (x = 31; x>=0; x--)
  {
    if ((number & (1 << x)) != 0)
      printf("1");
    else
      printf("0");
  }
  printf("\n");
}

/**
   Flip the second bit of the input.
*/
int
flip2(int i)
{
  static int s = 1 << 30;
  return i^s;
}

unsigned int
ilog2(unsigned int j)
{
  unsigned int r = 0;
  while (j >>= 1)
    r++;
  return r;
}

unsigned int
isqrt(unsigned int j)
{
  return (unsigned int) sqrt((double) j);
}

void
note(char* msg)
{
  if (msg == NULL)
    fprintf(debug_file, "[%i] (NULL) \n", debug_rank);
  else
    fprintf(debug_file, "[%i] %s \n", debug_rank, msg);

  fflush(debug_file);
}

void
note_b(char* msg, int b)
{
  fprintf(debug_file, "[%i] %s", debug_rank, msg);
  printbinary((unsigned int) b);
  fflush(debug_file);
}

void
note_d(char* msg, double d)
{
  fprintf(debug_file, "[%i] %s %g\n", debug_rank, msg, d);
  fflush(debug_file);
}

void
note_p(char* msg, void* p)
{
  fprintf(debug_file, "[%i] %s%p\n", debug_rank, msg, p);
  fflush(debug_file);
}

void
note_i(char* msg, int i)
{
  fprintf(debug_file, "[%i] %s%i\n", debug_rank, msg, i);
  fflush(debug_file);
}

void
note_ii(char* msg, int i, int j)
{
  fprintf(debug_file, "[%i] %s(%i,%i)\n", debug_rank, msg, i, j);
  fflush(debug_file);
}

void
note_u(char* msg, unsigned int u)
{
  fprintf(debug_file, "[%i] %s%ui\n", debug_rank, msg, u);
  fflush(debug_file);
}

void
note_xi(char* msg, int i, int j)
{
  fprintf(debug_file, "[%i] %s%X %i\n", debug_rank, msg, i, j);
  fflush(debug_file);
}

void
note_s(char* msg, char* s)
{
  assert(debug_file);
  fprintf(debug_file, "[%i] %s%s\n", debug_rank, msg, s);
  fflush(debug_file);
}

void
note_sx(char* msg, char* s, int i)
{
  fprintf(debug_file, "[%i] %s%s %X\n", debug_rank, msg, s, i);
  fflush(debug_file);
}

void
note_sxi(char* msg, char* s, int x, int i)
{
  fprintf(debug_file, "[%i] %s%s %X %i\n", debug_rank, msg, s, x, i);
  fflush(debug_file);
}

void
note_ss(char* msg, char* s1, char* s2)
{
  fprintf(debug_file, "[%i] %s%s %s", debug_rank, msg, s1, s2);
  fflush(debug_file);
}


/*
void
note_ssxi(char* msg, char* s1, char* s2, int x, int i)
{
  fprintf(debug_file, "[%i] %s%s %X %i\n", debug_rank, msg, s, x, i);
  fflush(debug_file);
}
*/

void
note_x(char* msg, int i)
{
  fprintf(debug_file, "[%i] %s%X\n", debug_rank, msg, i);
  fflush(debug_file);
}

void
note_xx(char* msg, int i, int j)
{
  fprintf(debug_file, "[%i] %s%X\n", debug_rank, msg, i);
  fflush(debug_file);
}

void
show_d(char* tag, double d)
{
  fprintf(debug_file, "[%i] %s: %g\n", debug_rank, tag, d);
  fflush(debug_file);
}

/*
void
show_i(char* tag, int i)
{
  fprintf(debug_file, "[%i] %s: %i\n", debug_rank, tag, i);
  fflush(debug_file);
}
*/

void
show_ii(char* tag_i, int i, char* tag_j, int j)
{
  fprintf(debug_file, "[%i] %s: %i, %s: %i\n", debug_rank, tag_i, i, tag_j, j);
  fflush(debug_file);
}

void
show_l(char* tag, long l)
{
  fprintf(debug_file, "[%i] %s: %li\n", debug_rank, tag, l);
  fflush(debug_file);
}

void
show_u(char* tag, unsigned int u)
{
  fprintf(debug_file, "[%i] %s: %u\n", debug_rank, tag, u);
  fflush(debug_file);
}

/*
void
show_s(char* tag, char* s)
{
  fprintf(debug_file, "[%i] %s: %s\n", debug_rank, tag, s);
  fflush(debug_file);
}
*/

void
show_x(char* tag, int x)
{
  fprintf(debug_file, "[%i] %s: %X\n", debug_rank, tag, x);
  fflush(debug_file);
}

void
show_b(char* tag, int b)
{
  fprintf(debug_file, "[%i] %s is %s\n", debug_rank, tag,
         b ? "true" : "false");
  fflush(debug_file);
}

void
show_p(char* tag, void* p)
{
  fprintf(debug_file, "[%i] %s: %p\n", debug_rank, tag, p);
  fflush(debug_file);
  fflush(stdout);
}

void
timestamp(char* label, char* msg)
{
  fprintf(debug_file, "%s", timestring(label, msg));
}

/**
   @return Formatted time string in static output location.
*/
char*
timestring(char* label, char* msg)
{
  static struct timeval tp;
  gettimeofday(&tp, NULL);
  if (msg)
    sprintf(output, "%s: at %li.%li %s \n",
            label, tp.tv_sec, tp.tv_usec, msg);
  else
    sprintf(output, "%s: at %li.%li \n",
           label, tp.tv_sec, tp.tv_usec);
  return output;
}

int
roll(int n)
{
  return (int) (n * 1.0 *
                (rand() / (RAND_MAX+1.0)));
}

/**
   Copy the string onto the heap so it can be used asynchronously.
   Just like strdup()
   @return A copy of the string or NULL if given NULL.
*/
char*
heap(char* s)
{
  // int length;
  char* h;
  if (s == NULL)
    return NULL;
  h = malloc((strlen(s)+1)*sizeof(char));
  assert(h != NULL);
  strcpy(h, s);
  return h;
}

/**
   Copy the int onto the heap as a decimal string.
*/
char*
iheap(int i)
{
  char* h = (char*) malloc(20*sizeof(char));
  assert(h != NULL);
  sprintf(h, "%i", i);
  return h;
}

/**
   Copy the int onto the heap as a hex string.
*/
char*
xheap(int i)
{
  char* h = (char*) malloc(20*sizeof(char));
  assert(h != NULL);
  sprintf(h, "%X", i);
  return h;
}

/**
   Return a random number in [0,n).
*/
int
rand_lt(int n)
{
  double d = (n*((double) rand())) / RAND_MAX;
  int result = (int) d;
  return result;
}

/**
   Return a random double in [0,n).
*/
double
random_double(int n)
{
  double d = (n*((double) rand())) / RAND_MAX;
  return d;
}

/**
   Read in a whole small file as a string.
*/
char*
file_string(char* file)
{
  intmax_t size;
  struct stat stats;

  stat(file, &stats);
  size = (intmax_t) stats.st_size;
  char* result = malloc(size+1);

  FILE* f = fopen(file, "r");
  int r = fread(result, 1, size, f);
  assert(r == size);
  result[size] = '\0';
  fclose(f);
  return result;
}

#ifdef ENABLE_DEBUG
#ifdef HAVE_EXEC_INFO_H
/**
   Dump the stack and exit.
*/
void
dump_stack()
{
  static void *buffer[100];
  // char **strings;
  int nptrs;
  nptrs = backtrace(buffer, 100);
  fprintf(debug_file, "backtrace() returned %d addresses\n", nptrs);
  fflush(NULL);
  backtrace_symbols_fd(buffer, nptrs, 1);
  exit(1);
}
#else
void
dump_stack()
{
  fprintf(debug_file, "cannot dump stack. \n");
}
#endif
#endif

#ifdef DMALLOC
/**
   Set up DMALLOC output
*/
void
dmalloc_setup(void)
{
  dmalloc_logpath = malloc(100);
  assert(dmalloc_logpath != NULL);
  sprintf(dmalloc_logpath, "./dmalloc_%i.out", debug_rank);
  SHOW_S(dmalloc_logpath);
  fflush(NULL);
}
#endif
