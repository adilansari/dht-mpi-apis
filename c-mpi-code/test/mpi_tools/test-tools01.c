
/**
   Test math functions and simple output.
*/

#include <mpi_tools.h>

int
main(int argc, char* argv[])
{
  MPI_Init(&argc, &argv);

  whoami();
  setup_debug_file();

  int i;
  for (i = 0; i < 20; i++)
  {
    SHOW_II(i, ilog2(i));
    SHOW_II(i, isqrt(i));
  }

  MPI_Finalize();

  return 0;
}
