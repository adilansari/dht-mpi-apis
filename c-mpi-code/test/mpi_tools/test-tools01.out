Debug to stdout
[0] i: 0, ilog2(i): 0
[0] i: 0, isqrt(i): 0
[0] i: 1, ilog2(i): 0
[0] i: 1, isqrt(i): 1
[0] i: 2, ilog2(i): 1
[0] i: 2, isqrt(i): 1
[0] i: 3, ilog2(i): 1
[0] i: 3, isqrt(i): 1
[0] i: 4, ilog2(i): 2
[0] i: 4, isqrt(i): 2
[0] i: 5, ilog2(i): 2
[0] i: 5, isqrt(i): 2
[0] i: 6, ilog2(i): 2
[0] i: 6, isqrt(i): 2
[0] i: 7, ilog2(i): 2
[0] i: 7, isqrt(i): 2
[0] i: 8, ilog2(i): 3
[0] i: 8, isqrt(i): 2
[0] i: 9, ilog2(i): 3
[0] i: 9, isqrt(i): 3
[0] i: 10, ilog2(i): 3
[0] i: 10, isqrt(i): 3
[0] i: 11, ilog2(i): 3
[0] i: 11, isqrt(i): 3
[0] i: 12, ilog2(i): 3
[0] i: 12, isqrt(i): 3
[0] i: 13, ilog2(i): 3
[0] i: 13, isqrt(i): 3
[0] i: 14, ilog2(i): 3
[0] i: 14, isqrt(i): 3
[0] i: 15, ilog2(i): 3
[0] i: 15, isqrt(i): 3
[0] i: 16, ilog2(i): 4
[0] i: 16, isqrt(i): 4
[0] i: 17, ilog2(i): 4
[0] i: 17, isqrt(i): 4
[0] i: 18, ilog2(i): 4
[0] i: 18, isqrt(i): 4
[0] i: 19, ilog2(i): 4
[0] i: 19, isqrt(i): 4
